core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTypeUtil.java

```diff
@@ -82,6 +99,73 @@
                     .put("0", false)
                     .build();
 
+    /**
+     * Try to get a public static method {@link MethodHandle} with a single {@link String} argument
+     * in clazz.
+     * @param <T> the expected result
+     * @param clazz the class being introspected
+     * @param methodName the method expected
+     * @return a function that takes a {@link String} and produces a {@link T}
+     */
+    @Nullable
+    private static <T> MethodHandle getPublicStaticMethodHandle(final Class<T> clazz,
+                                                                final String methodName) {
+        try {
+            final MethodType methodType = MethodType.methodType(clazz, String.class);
+            return MethodHandles.publicLookup().findStatic(clazz, methodName, methodType);
+        } catch (Throwable t) {
+            // No valid public static method found
+            return null;
+        }
+    }
+
+    /**
+     * Try to get a constructor {@link MethodHandle} with a single {@link String} argument in clazz.
+     * @param <T> the expected result
+     * @param clazz the class being introspected
+     * @return a function that takes a {@link String} and produces a {@link T}
+     */
+    @Nullable
+    private static <T> MethodHandle getStringConstructorMethodHandle(final Class<T> clazz) {
+        try {
+            final MethodType methodType = MethodType.methodType(void.class, String.class);
+            return MethodHandles.publicLookup().findConstructor(clazz, methodType);
+        } catch (Throwable t) {
+            // No valid constructor found
+            return null;
+        }
+    }
+
+    /**
+     * Try to get a function that can create an instance of type {@code clazz} with a single
+     * {@link String} argument.
+     * @param <T> the expected resulting type
+     * @param clazz the class being introspected
+     * @return a function that takes a {@link String} and produces a {@code T}
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    static <T> Function<String, T> getCreatorMethod(Class<T> clazz) {
+        final MethodHandle methodHandle = Stream.of("of", "valueOf", "fromString")
+                     .map((methodName) -> getPublicStaticMethodHandle(clazz, methodName))
+                     .filter(Objects::nonNull)
+                     .findFirst()
+                     .orElseGet(() -> getStringConstructorMethodHandle(clazz));
+        if (methodHandle != null) {
```

Comment:
How about returning early when `methodHandle == null` so we reduce the cognitive load of the reader?

Reply:
Done!

I've gotten too used to Kotlin and `if`-is-expression, where this makes sense (single return or even no explicit return statement in the method).